src/function/maybe_changed_after.rs

```diff
@@ -1,14 +1,45 @@
 use crate::{
     accumulator::accumulated_map::InputAccumulatedValues,
-    ingredient::MaybeChangedAfter,
+    cycle::CycleRecoveryStrategy,
     key::DatabaseKeyIndex,
+    table::sync::ClaimResult,
     zalsa::{Zalsa, ZalsaDatabase},
     zalsa_local::{ActiveQueryGuard, QueryEdge, QueryOrigin},
     AsDynDatabase as _, Id, Revision,
 };
+use rustc_hash::FxHashSet;
 
 use super::{memo::Memo, Configuration, IngredientImpl};
 
+/// Result of memo validation.
+pub enum VerifyResult {
+    /// Memo has changed and needs to be recomputed.
+    Changed,
+
+    /// Memo remains valid.
+    ///
+    /// The first inner value tracks whether the memo or any of its dependencies have an
+    /// accumulated value.
+    ///
+    /// Database keys in the hashset represent cycle heads encountered in validation; don't mark
+    /// memos verified until we've iterated the full cycle to ensure no inputs changed.
+    Unchanged(InputAccumulatedValues, FxHashSet<DatabaseKeyIndex>),
```

Comment:
As discussed in our sync meeting. It might be worthwhile to use an `Option<Box<FxHashSet<...>>>` here to reduce the size of `VerifyResult`. It would be even better if we could store the cycle heads somewhere else to avoid `VerifyResult` implementing `Drop` but that could be tricky and might not be worth it just to safe one extra condition.

Reply:
Ah, Micha already proposed a wrapper type a couple reviews below https://github.com/salsa-rs/salsa/pull/603#discussion_r1946252725

Reply:
Done